src/cargo/ops/cargo_config.rs

```diff
@@ -4,20 +4,35 @@ use crate::util::config::{Config, ConfigKey, ConfigValue as CV, Definition};
 use crate::util::errors::CargoResult;
 use crate::{drop_eprintln, drop_println};
 use anyhow::{bail, format_err, Error};
+use clap::{ArgEnum, PossibleValue};
 use serde_json::json;
 use std::borrow::Cow;
 use std::fmt;
 use std::str::FromStr;
 
+#[derive(clap::ArgEnum, Clone)]
 pub enum ConfigFormat {
     Toml,
     Json,
     JsonValue,
 }
 
 impl ConfigFormat {
-    /// For clap.
-    pub const POSSIBLE_VALUES: &'static [&'static str] = &["toml", "json", "json-value"];
+    pub fn possible_values() -> impl Iterator<Item = PossibleValue<'static>> {
+        ConfigFormat::value_variants()
+            .iter()
+            .filter_map(ArgEnum::to_possible_value)
+    }
+}
+
+impl AsRef<str> for ConfigFormat {
+    fn as_ref(&self) -> &str {
+        match self {
+            ConfigFormat::Toml => "toml",
+            ConfigFormat::Json => "json",
+            ConfigFormat::JsonValue => "json-value",
+        }
+    }
 }
```

Comment:
I think you can use `PossibleValue` to help with this, like we do `Display` in `clap_complete::Shell`.

```rust
impl Display for Shell {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.to_possible_value()
            .expect("no values are skipped")
            .get_name()
            .fmt(f)
    }
}
```

Reply:
for my education, what's the difference between using a `PossibleValue` vs creating a match arm that returns static strings? Are both of them exhaustive checks?

Reply:
This single-sources the values so you don't have to worry about typos.

Reply:
To confirm, should I replace the derived clap::ArgEnum implementation with my own impl that has `to_possible_value()` and `value_variants()` implementations?

```bash
error[E0119]: conflicting implementations of trait `clap::ArgEnum` for type `ops::cargo_config::ConfigFormat`
  --> src/cargo/ops/cargo_config.rs:13:10
   |
13 | #[derive(clap::ArgEnum, Clone)]
   |          ^^^^^^^^^^^^^ conflicting implementation for `ops::cargo_config::ConfigFormat`
...
28 | impl ArgEnum for ConfigFormat {
   | ----------------------------- first implementation here
```

Reply:
No, you can keep deriving, just use the functions from the trait to get the generated values.  `Shell` doesn't derive to avoid foisting the `derive` feature on people not using it.

Reply:
thanks for the tip and sorry if I am being daft here. 

```rust
        self.to_possible_value()
            .expect("no values are skipped")
            .get_name()
```

this looks like a runtime check that might throw at runtime, while the current implementation is an exhaustive compile-time check. Should I still use the `to_possible_value().expect("Should be fine")` and if so, what's the safest way to do that?

Reply:
`to_possible_value` only "fails" if you have a `#[clap(skip)]` field.  None of them do which is why an `expect` is ok here to act as an assert (the `expect` is documenting the invariant that it is relying on).

This is better because the other approach would have required tests for each field to make sure there were no typos.